impeller/display_list/dl_aiks_canvas.cc

```diff
@@ -0,0 +1,1191 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/display_list/dl_aiks_canvas.h"
+
+#include <algorithm>
+#include <cstring>
+#include <memory>
+#include <optional>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/trace_event.h"
+#include "impeller/aiks/color_filter.h"
+#include "impeller/core/formats.h"
+#include "impeller/display_list/dl_image_impeller.h"
+#include "impeller/display_list/dl_vertices_geometry.h"
+#include "impeller/display_list/nine_patch_converter.h"
+#include "impeller/display_list/skia_conversions.h"
+#include "impeller/entity/contents/conical_gradient_contents.h"
+#include "impeller/entity/contents/filters/filter_contents.h"
+#include "impeller/entity/contents/filters/inputs/filter_input.h"
+#include "impeller/entity/contents/linear_gradient_contents.h"
+#include "impeller/entity/contents/radial_gradient_contents.h"
+#include "impeller/entity/contents/runtime_effect_contents.h"
+#include "impeller/entity/contents/scene_contents.h"
+#include "impeller/entity/contents/sweep_gradient_contents.h"
+#include "impeller/entity/contents/tiled_texture_contents.h"
+#include "impeller/entity/entity.h"
+#include "impeller/entity/geometry/geometry.h"
+#include "impeller/geometry/path.h"
+#include "impeller/geometry/path_builder.h"
+#include "impeller/geometry/scalar.h"
+#include "impeller/geometry/sigma.h"
+#include "impeller/typographer/backends/skia/text_frame_skia.h"
+
+namespace impeller {
+
+#define UNIMPLEMENTED \
+  FML_DLOG(ERROR) << "Unimplemented detail in " << __FUNCTION__;
+
+DlAiksCanvas::DlAiksCanvas(const SkRect& cull_rect, bool prepare_rtree)
```

Comment:
This is all quite a bit different from what I thought was being done.

As originally described to me, I had thought that Impeller would continue to just have the dispatcher, but the conversion of DlCanvas to DlOpReceiver calls was going to happen independently and then that conversion would just talk to DlDispatcher directly.

As it happens, there are a few mistakes in this and duplication of code (or failure to duplicate important code) from DisplayListBuilder in addition to basically duplicating all of impeller's DlDispatcher.

What I though was going to happen was:

- currently DisplayListBuilder talks DlCanvas and optimizes the calls and then records them into a DL buffer it takes "DlCanvas" in, but it records "DlOpReceiver" calls on the back end
- Take the part of the Builder that does the optimizations and pull it out into a DisplayListCanvasOptimizer class that forwards the calls to a backend which implements DlOpReceiver
- Take the part of the Builder that records the Op calls into a buffer and pull that out into a DisplayListRecorder class that just records the calls into a buffer.
- DisplayListBuilder would simply link up a DLOptimizer with a DlRecorder and maintain the same functionality.
- For Impeller, DLOptimizer would talk to the existing impeller::DlDispatcher and it would get all of the optimizations for free without having to reinvent the wheel or create essentially a line-for-line replacement of DlDispatcher with, as we see here, missing optimizations and a couple of bugs.
- That's what I thought the effort was going to be, but I see now that I didn't hear that right.